src/algebra/homology/local_cohomology.lean

```diff
@@ -0,0 +1,173 @@
+* Establish long exact sequence(s) in local cohomology
+-/
+
+open opposite
+open category_theory
+open category_theory.limits
+
+noncomputable theory
+
+universes u v
+
+/- We define local cohomology, implemented as a direct limit of Ext(R/J, --). -/
+section local_cohomology
+
+section
+variables {R : Type u} [comm_ring R] {D : Type v} [small_category D]
+
+local attribute [ext] quiver.hom.unop_inj
+
+/--  The directed system of `R`-modules of the form `R/J`, where `J` is an ideal of `R`,
+determined by the functor `I`  -/
+def ring_mod_ideals (I : D ⥤ ideal R) : D ⥤ Module.{u} R :=
+{ obj := λ t, Module.of R $ R ⧸ (I.obj t),
+  map := λ s t w, submodule.mapq _ _ (linear_map.id) (I.map w).down.down }
+
+/-- The diagram we will take the colimit of to define local cohomology, corresponding to the
+directed system determined by the functor `I` -/
+def local_cohomology_diagram (I : D ⥤ ideal R) (i : ℕ) : Dᵒᵖ ⥤ Module.{u} R ⥤ Module.{u} R :=
+(ring_mod_ideals I).op ⋙ Ext R (Module.{u} R) i
+
+end
+
+section
+-- We momentarily need to work with a type inequality, as later we will take colimits
+-- along diagrams either in Type, or in the same universe as the ring, and we need to cover both.
+variables {R : Type max u v} [comm_ring R] {D : Type v} [small_category D]
+
+/-- `local_cohomology I i` is the the functor sending an module `M` over a commutative ring `R`
+to the direct limit of `Ext^i(R/J, M)`, where `J` ranges over a collection of ideals of `R`
+that is represented as a functor `I`.
+
+In this definition we do not assume any special property of the diagram `I`, but the relevant case
+will be where `I` is cofinal with the diagram of powers of a single given ideal.
+
+Below, we give two equivalent (to be shown) definitions of the usual local cohomology with support
+in an ideal `J`, `local_cohomology_powers` and `local_cohomology_univ`.
+
+TODO: Show that any functor cofinal with `I` gives the same result.
+ -/
+def local_cohomology (I : D ⥤ ideal R) (i : ℕ) :
+  Module.{max u v} R ⥤ Module.{max u v} R :=
+colimit (local_cohomology_diagram.{(max u v) v} I i)
+
+end
+
+/- We give two models for the local cohomology with support in an ideal `J`: first in terms of
+the powers of `J` (`local_cohomology_powers`), then in terms of all ideals with the same radical
+as `J` (`local_cohomology_univ`). -/
+section models_for_local_cohomology
+
+variables {R : Type u} [comm_ring R]
+
+/-- The functor sending a natural number `i` to the `i`-th power of the ideal `J` -/
+def ideal_powers (J : ideal R) : ℕᵒᵖ ⥤ ideal R :=
+{ obj := λ t, J^(unop t),
+  map := λ s t w, ⟨⟨ideal.pow_le_pow w.unop.down.down⟩⟩, }
+
+/-- `local_cohomology_powers J i` is `i`-th the local cohomology module of a module `M` over
+a commutative ring `R` with support in the ideal `J` of `R`, defined as the direct limit
+of Ext(R/J^t, M) over all powers `t`. -/
+def local_cohomology_powers (J : ideal R) (i : ℕ) : Module.{u} R ⥤ Module.{u} R :=
```

Comment:
I think this is the one which should be called local cohomology because it's what people mean by the term.